### DISCUSSION TITLE
Nesting - particles stop moving when leaving the fine resolution grid

### DISCUSSION THREAD

**Discussion by cgrimaldi** [Parcels-code/Parcels#1674] [Q&A]
## Question
Hi 
I am trying to use the nesting function in ocean parcels, but can't seem to get it to work. I do not get any error, however when I look at my particle tracks output, it stops moving when it leaves my finer grid resolution model. I am assuming I am doing something wrong in the nesting side of things, but have followed the tutorial pretty closely. See code below. Any help is appreciated thanks :)

```python

import matplotlib.pyplot as plt
import numpy as np
import xarray as xr
import pandas
from datetime import timedelta as delta
from datetime import datetime as datetime

from parcels import AdvectionRK4, Field, FieldSet, JITParticle, NestedField, ParticleSet

def DeleteParticle(particle, fieldset, time):
    particle.delete()


# LOAD COARSER RESOLUTION GRID - ROMS
fname = 'ROMS.nc'
filenames = {'U': fname, 'V': fname}
variables = {'U': 'u_eastward', 'V': 'v_northward'}
dimensions = {'U': { 'lon': 'lon_rho_utm','lat': 'lat_rho_utm','time': 'ocean_time'},
              'V': { 'lon': 'lon_rho_utm','lat': 'lat_rho_utm','time': 'ocean_time'}}
fieldset_Large = FieldSet.from_netcdf(filenames, variables, dimensions, mesh="flat",allow_time_extrapolation = True)

# PLOT COARSER RESOLUTION GRID - ROMS
fieldset_Large.computeTimeChunk()
plt.pcolormesh(fieldset_Large.U.grid.lon, fieldset_Large.U.grid.lat, fieldset_Large.U.data[0,:,:])
plt.xlabel("Northings[m]")
plt.ylabel("Eastings[m]")
plt.colorbar()
plt.show()

# READ FINER RESOLUTION GRID - DFM
fname='DFM.nc'
filenames = {'U': fname, 'V': fname}
variables = {'U': 'U', 'V': 'V'}
dimensions = {'U': {'lon': 'x','lat': 'y','time': 'time'},'V': {'lon': 'x','lat': 'y','time': 'time'}}
fieldset_Small = FieldSet.from_netcdf(filenames, variables, dimensions, mesh="flat",allow_time_extrapolation = True)

# PLOT FINER RESOLUTION GRID - DFM
fieldset_Small.computeTimeChunk()
plt.pcolormesh(fieldset_Small.U.grid.lon, fieldset_Small.U.grid.lat, fieldset_Small.U.data[1,:, :])
plt.xlabel("Northings[m]")
plt.ylabel("Eastings[m]")
plt.colorbar()
plt.show()

# CREATE NESTED FIELD
U = NestedField('U', [fieldset_Small.U,fieldset_Large.U])
V = NestedField('V', [fieldset_Small.V,fieldset_Large.V])
fieldset_nest = FieldSet(U, V)

# RUN PTRACKING
pset = ParticleSet(fieldset_nest, pclass=JITParticle,lon=[200000], lat=[7590000])

output_file = pset.ParticleFile(
    name="NestedFieldParticle_test.zarr", outputdt=delta(days=1))

pset.execute(AdvectionRK4, runtime=delta(days=30), dt=delta(days=1), output_file=output_file)

```

**Comment by VeckoTheGecko** [comment]
Hi Camille!

There doesn't seem to be anything wrong with the code that you are working with. Would you be able to attach the plotted trajectories as well?

Cheers,
Nick

**Reply by cgrimaldi** [reply]
Hi Nick
Yes see attached, just plotting one particle here and it goes around fine until it reaches the end of the fine-scale grid, when it just stagnates; the lon, lat at the end of the files are all within a couple of m from each other.
Thanks for your help!
Camille
![example](https://github.com/user-attachments/assets/0a7bf6e2-df48-48a4-8c2d-8e6a133350ac)

**Reply by VeckoTheGecko** [reply]
I'd like to get to the bottom of this, and the first step would be to get a reproducible example of this on my machine.

How big is the data that you're working with @cgrimaldi ? Do you think it would be possible to subset the data (e.g., take one time slice and add `allow_time_extrapolation=True`, and restrict the spatial domain) down to something I can download, make sure its still reproducible, zip the code and data together and provide it to me via [wetransfer](https://wetransfer.com/) or something similar?

**Reply by cgrimaldi** [reply]
Hi!
Thanks so much for looking into it - have tried multiple different things and troubleshooting with people from my lab but we couldn't get to the bottom of it. I have uploaded the files here: https://we.tl/t-2ZfzDbkLRH. Let me know if you find something!
Cheers,
Camille

**Reply by VeckoTheGecko** [reply]
I think the problem is that there are nans in the smaller domain in the northern region which connects to the larger model. These nans are treated as 0s in Parcels (i.e., land). This is different to something leaving the field and giving a particle out of bounds error (which is how the NestedField knows to check the other field). I think having support for non-rectangular inner-nested fields would require a major rework in Parcels, adding support for non-rectangular fields.

```py
ds = xr.open_dataset('DFM_small.nc')
(ds.U.isel(time=-1).isnull()).plot()
```

![output](https://github.com/user-attachments/assets/8b4f4f28-41a6-4251-bd46-dcdbb63d107d)

Unfortunately the only idea I have on how to proceed would be to do some preprocessing to fill those nans.

@erikvansebille any ideas on how to proceed?

**Reply by michaeldenes** [reply]
Perhaps this is a silly suggestion, but could you ignore the nested fields component, and write a custom kernel for advection. Creating a fieldset with vector fields '`(Small_U,SmallV)`, and `(Large_U, Large_V)`, something along the lines of:
```python
(u_small,v_small) = fieldset.small_UV[particle]

if np.sqrt(u**2+v**2) < 1e-12:
    (u,v) = fieldset.large_UV[particle]
else:
    (u,v) = (u_small,v_small)
```
and then proceed as per usual (and doing something slightly smarter, of course)?

**Reply by erikvansebille** [reply]
Following on from @michaeldenes' suggestion, an alternative that still sues the NestedFields implementation is to create a Kernel that raises an OutOfBoundsError when the local velocities are zero

Something like (untested):
```python
def CheckInSmallDomain(particle, fieldset, time):
    u,v = fieldset.UV[time, particle.depth, particle.lat, particle.lon]
    if (u**2+v**2) < 1e-12:
        particle.state = StatusCode.ErrorOutOfBounds
```

**Reply by cgrimaldi** [reply]
Hi all
Sorry for the late reply, thanks for all the suggestions.

I see @VeckoTheGecko, thanks for looking into it. Tricky if the problem comes from the non rectangular grid.

@michaeldenes @erikvansebille, I am pretty new at writing kernels so sorry if I am not the most efficient. I have tried your suggestions but I am getting the following error: 

NotImplementedError: Cannot convert numpy functions in kernels to C-code.
Either use functions from the math library or run Parcels in Scipy mode.
For more information, see https://docs.oceanparcels.org/en/latest/examples/tutorial_parcels_structure.html#3.-Kernels

Any idea how to get around that?
Thanks
Camille

**Reply by erikvansebille** [reply]
Oh sorry, my mistake. You should not use the `np.sqrt()`, as `numpy` functions can't be used in Kernels. I edited the proposed kernel above

**Reply by cgrimaldi** [reply]
Oh right - easy fix thanks @erikvansebille. From what I understand, here you are just creating a kernel to check when the particle is in the small domain, and if it is, you pop out an out of bond error, but how do I actually make it use the larger grid velocities when it leaves the small grid?

I was trying to do something like that, based on my code higher up, does it make sense? 
```py
def CheckInSmallDomain(particle, fieldset, time):
    u,v = fieldset.UV[time, particle.depth, particle.lat, particle.lon]
    if (u**2+v**2) < 1e-12:
        (u,v) = fieldset_Large.UV[particle]
    else:
        (u,v) = fieldset_Small.UV[particle]

% RUN PTRACKING
pset = ParticleSet(fieldset_Small, pclass=JITParticle,lon=[200000], lat=[7630000])
output_file = pset.ParticleFile(name="NestedFieldParticle_test.zarr", outputdt=delta(days=1))
pset.execute([AdvectionRK4,CheckInSmallDomain], runtime=delta(days=30), dt=delta(days=1), output_file=output_file)
```
But getting this error: `NotImplementedError: Cannot convert 'fieldset_Large' used in kernel to C-code`

**Reply by erikvansebille** [reply]
Well, that is the point of the `NestedFields`; under the hood Parcels, will go to the next Field in the list of Fields when an OutofBounds error occurs; so no need to do anything with fieldset_large and fieldset_small etc. I hope that with just my proposed `CheckInSmallDomain()` kernel above, your nesting will work as you implement it at the top

**Reply by cgrimaldi** [reply]
I was going from the suggestion that @michaeldenes made about going around the nested field. 
When I use your suggestion, it just raises an out of bond error when the particle leaves the smaller domain..

def CheckInSmallDomain(particle, fieldset, time):
    u,v = fieldset.UV[time, particle.depth, particle.lat, particle.lon]
    if (u**2+v**2) < 1e-12:
        particle.state = StatusCode.ErrorOutOfBounds

**Reply by cgrimaldi** [reply]
@VeckoTheGecko it seems like if I manually modify the NaN with a random missing value in the netcdf file, the nested field works. I will let a couple of simulations run and report back tomorrow. Hopefully that works, thanks for the NaN hint!
Camille
![image](https://github.com/user-attachments/assets/3db08212-374e-439c-a7db-d01c63d6d5b4)
